gsplat/rendering.py

```diff
@@ -560,3 +571,218 @@ def _getProjectionMatrix(znear, zfar, fovX, fovY, device="cuda"):
+        camera_ids=camera_ids,
+        gaussian_ids=gaussian_ids,
+    )
+    isect_offsets = isect_offset_encode(isect_ids, C, tile_width, tile_height)
+
+    # TODO: SH also suport N-D.
+    # Compute the per-view colors
+    if not (
+        colors.dim() == 3 and sh_degree is None
+    ):  # silently support [C, N, D] color.
+        colors = (
+            colors[gaussian_ids] if packed else colors.expand(C, *([-1] * colors.dim()))
+        )  # [nnz, D] or [C, N, 3]
+    else:
+        if packed:
+            colors = colors[camera_ids, gaussian_ids, :]
+    if sh_degree is not None:  # SH coefficients
+        camtoworlds = torch.inverse(viewmats)
+        if packed:
+            dirs = means[gaussian_ids, :] - camtoworlds[camera_ids, :3, 3]
+        else:
+            dirs = means[None, :, :] - camtoworlds[:, None, :3, 3]
+        colors = spherical_harmonics(
+            sh_degree, dirs, colors, masks=radii > 0
+        )  # [nnz, D] or [C, N, 3]
+        # make it apple-to-apple with Inria's CUDA Backend.
+        colors = torch.clamp_min(colors + 0.5, 0.0)
+
+    # Rasterize to pixels
+    if render_mode in ["RGB+D", "RGB+ED"]:
+        colors = torch.cat((colors, depths[..., None]), dim=-1)
+    elif render_mode in ["D", "ED"]:
+        colors = depths[..., None]
+    else:  # RGB
+        pass
+    densifications = (
+        torch.zeros_like(
+            means2d, dtype=means2d.dtype, requires_grad=True, device="cuda"
+        )
+        + 0
+    )
+    render_colors, render_alphas, render_normals, render_distloss = rasterize_to_pixels_2dgs(
+        means2d,
+        densifications,
+        ray_transformations,
+        colors,
+        opacities,
+        normals,
+        width,
+        height,
+        tile_size,
+        isect_offsets,
+        flatten_ids,
+        backgrounds=backgrounds,
+        packed=packed,
+        absgrad=absgrad,
+        distloss=distloss,
+    )
+    render_normals_from_depth = None
+    if render_mode in ["ED", "RGB+ED"]:
+        # normalize the accumulated depth to get the expected depth
+        render_colors = torch.cat(
+            [
+                render_colors[..., :-1],
+                render_colors[..., -1:] / render_alphas.clamp(min=1e-10),
+            ],
+            dim=-1,
+        )  
+    if render_mode in ["RGB+ED", "RGB+D"]:
+        render_depths = render_colors[..., -1:]
+        render_normals_from_depth = depth_to_normal(
+            render_depths,
+            torch.linalg.inv(viewmats), # c2w 
+            Ks, 
+            near_plane,
+            far_plane
+        ).squeeze(0)
+
+        # render_normals_from_depth = depth_to_normal(
+        #     torch.linalg.inv(viewmats.squeeze(0)),
+        #     Ks.squeeze(0),
+        #     width,
+        #     height,
+        #     render_depths,
+        # )
+    
+    meta = {
+        "camera_ids": camera_ids,
+        "gaussian_ids": gaussian_ids,
+        "radii": radii,
+        "means2d": densifications,
```

Comment:
Why do we use densification instead of means2d here?

Reply:
Hey, sorry for the very late reply. 
The densifications is actually a dummy variable that holds the gradients for densification. You can find the corresponding cuda code here: https://github.com/nerfstudio-project/gsplat/blob/56d0a81c089966441168cc66380bf2ec4732be77/gsplat/cuda/csrc/projection.cu#L2120-L2122